src/zip.c

```diff
@@ -1038,3 +1074,395 @@ struct zip_t *zip_open_stream(const char *stream, size_t size) {
+    } else {
+      entry_mark_array[i].type = KEEP;
+    }
+
+    CLEANUP(zip->entry.name);
+
+    if (!mz_zip_reader_file_stat(&zip->archive, i, &file_stat)) {
+      return -1;
+    }
+
+    entry_mark_array[i].m_local_header_ofs = file_stat.m_local_header_ofs;
+    entry_mark_array[i].file_index = -1;
+    entry_mark_array[i].lf_length = 0;
+    if ((entry_mark_array[i].type) == DELETE &&
+        (d_pos > entry_mark_array[i].m_local_header_ofs)) {
+      d_pos = entry_mark_array[i].m_local_header_ofs;
+    }
+  }
+  for (int i = 0; i < n; ++i) {
+    if ((entry_mark_array[i].m_local_header_ofs > d_pos) &&
+        (entry_mark_array[i].type != DELETE)) {
+      entry_mark_array[i].type = MOVE;
+    }
+  }
+  return 0;
+}
+
+static int get_new_index(mz_uint64 *local_header_ofs_array, int cur_index) {
+  int new_index = 0;
+  for (int i = cur_index - 1; i >= 0; --i) {
+    if (local_header_ofs_array[cur_index] > local_header_ofs_array[i]) {
+      new_index = i + 1;
+      return new_index;
+    }
+  }
+  return new_index;
+}
+
+static int sort_insert(mz_uint64 *local_header_ofs_array, int cur_index) {
+  int new_index = get_new_index(local_header_ofs_array, cur_index);
+
+  if (new_index != cur_index) {
+    mz_uint64 temp = local_header_ofs_array[cur_index];
+    for (int i = cur_index; i > new_index; i--) {
+      local_header_ofs_array[i] = local_header_ofs_array[i - 1];
+    }
+    local_header_ofs_array[new_index] = temp;
+  }
+  return new_index;
+}
+
+static int update_file_index_array(struct entry_mark *entry_mark_array,
+                                   int end_index, int insert_index) {
+  for (int j = 0; j < end_index; j++) {
+    if (entry_mark_array[j].file_index >= insert_index) {
+      entry_mark_array[j].file_index += 1;
+    }
+  }
+  entry_mark_array[insert_index].file_index = end_index;
+  return 0;
+}
+
+static int finalize_entry_mark_array(struct zip_t *zip,
+                                     struct entry_mark *entry_mark_array,
+                                     const int n) {
+
+  mz_uint64 *local_header_ofs_array = (mz_uint64 *)calloc(n, sizeof(mz_uint64));
+  if (local_header_ofs_array == NULL) {
+    return -1;
+  }
+
+  for (int i = 0; i < n; ++i) {
+    local_header_ofs_array[i] = entry_mark_array[i].m_local_header_ofs;
+    int new_index = sort_insert(local_header_ofs_array, i);
+
+    if (new_index != i) {
+      update_file_index_array(entry_mark_array, i, new_index);
+    }
+    entry_mark_array[i].file_index = new_index;
+  }
+
+  mz_uint64 *length = (mz_uint64 *)calloc(n, sizeof(mz_uint64));
+  if (length == NULL) {
+    CLEANUP(local_header_ofs_array);
+    return -1;
+  }
+  for (int i = 0; i < n - 1; i++) {
+    length[i] = local_header_ofs_array[i + 1] - local_header_ofs_array[i];
+  }
+  length[n - 1] = zip->archive.m_archive_size - local_header_ofs_array[n - 1];
+
+  for (int i = 0; i < n; i++) {
+    entry_mark_array[i].lf_length = length[entry_mark_array[i].file_index];
```

Comment:
Can we calculate it in one loop (above)?

Reply:
Most likely not. 
According to
[4.4.1.3  The entries in the central directory MAY NOT necessarily
      be in the same order that files appear in the .ZIP file.](https://pkware.cachefly.net/webdocs/casestudies/APPNOTE.TXT)
The order of length and entry_mark_array may be different.
so : entry_mark_array[i].lf_length = length[entry_mark_array[i].file_index];
It's not the same order.